llvm/lib/Target/AMDGPU/SMInstructions.td

```diff
@@ -29,7 +29,6 @@ class SM_Pseudo <string opName, dag outs, dag ins, string asmOps, list<dag> patt
   let mayStore = 0;
   let mayLoad = 1;
   let hasSideEffects = 0;
-  let maybeAtomic = 0;
```

Comment:
How does removing this cause SIMemoryLegalizer to process scalar loads and stores?

Should the memory model in AMDGPUUsage be updated to include the scalar memory instructions?

Reply:
> How does removing this cause SIMemoryLegalizer to process scalar loads and stores?

Not setting `maybeAtomic = 0` on SMEM Pseudos means that they all inherit the default of `maybeAtomic = 1`. The main loop in `SIMemoryLegalizer::runOnMachineFunction` ignores anything that is not marked as `maybeAtomic`.

> Should the memory model in AMDGPUUsage be updated to include the scalar memory instructions?

Probably. Are you saying that should be a prerequisite for this patch?

Reply:
Yes, I think if we change the implementation of the memory model we should ensure the ABI documentation is updated to match. I think only the non-atomic rows need to be updated to include the SMEM instructions since we do not support scalar atomics.